### PR TITLE
Working WMArchive performance prototype

### PR DESCRIPTION
**Pull request by nilsvu**
- Discussed in my [WMArchiveAggregation](https://github.com/knly/WMArchiveAggregation/) reports
- Server is running on `vocms013:8249`

### CODE REVIEW

**Comment by vkuznet** [comment]
comment

File: src/css/main.scss [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+$body-color: $text;
+$link-color: $highlight;
+$headings-color: $structure;
+$input-border-focus: $highlight;
+$input-border-color: $light;
+$input-color-placeholder: $light;
+$btn-secondary-border: $light;
+
+// Bootstrap
+@import "../dependencies/bootstrap/scss/bootstrap-flex.scss";
```
Comment (vkuznet) [inline src/css/main.scss]: Nils, I don't like relative path here. The problem is that during RPM build we relocate stuff, see
https://github.com/cms-sw/cmsdist/blob/comp_gcc493/wmarchive.spec#L50

and it may affect the import here. Please have a look at vocms013:/data/srv/HG1512e/sw.valya/slc6_amd64_gcc493/cms/wmarchive/v00.00.70
WMArchive dir structure once RPM are in place. It is different from package one. And, make adjustment to import here to make it work properly within a package and RPM build/layout.

File: src/dependencies/bootstrap/Gemfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Ruby Gems for building and testing Bootstrap
```
Comment (vkuznet) [inline src/dependencies/bootstrap/Gemfile]: What dependencies dir is doing? We will not run any bootstrap script, instead we use RPMs to install packages. If you need list of packages we'll need to create spec files for them and then build their RPMs. Please explain how you install stuff and how dependencies dir is working with its internals. This should be solved before I'll make a build.

File: src/dependencies/bootstrap/Gemfile [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+# Run `grunt update-gemfile-lock` to update to the latest compatible versions
+
+source 'https://rubygems.org'
```
Comment (vkuznet) [inline src/dependencies/bootstrap/Gemfile]: why do we need ruby? Do you need it to install dependencies? Please make ASCII list of dependencies and we should start writing spec files for them.

File: src/js/utils.js [hunk 4]
```diff
@@ -1,200 +0,0 @@
-/*
```
Comment (vkuznet) [inline src/js/utils.js:1]: Why it is removed? and how do you replace it?

**Comment by vkuznet** [comment]
Nils, am I correct that you included bootstrap package into src/dependencies? If so, could you please remove it completely. We'll build it separately. For that we'll need to create a spec file in cmsdist, create a bootstrap RPM and make WMArchvie spec depend on it. See for example: https://github.com/cms-sw/cmsdist/blob/comp_gcc493/wmarchive.spec#L9

**Comment by vkuznet** [comment]
This will clear up your PR. Please create bootstrap.spec file similar to any other spec you can find in cmsdist repository, e.g. https://github.com/cms-sw/cmsdist/blob/comp_gcc493/yui3.spec

Then identify what pieces we need from such RPM and only add them to WMArchive. So far it is hard for me to see in your PR where bootstrap is used.

**Comment by nilsvu** [comment]
Building bootstrap is not necessary. Bootstrap is a responsive CSS framework that is distributed as a SASS library. [SASS](http://sass-lang.com) is a CSS preprocessor: Every CSS file is also valid SASS syntax, but SASS adds _a lot_ of very important functionality to CSS such as variables and nested selectors. The `src/sass/main.scss` is compiled to the `src/css/main.css` file that is then a regular CSS file and included in `src/templates/index.html`. The usual workflow is to automatically compile the SASS file on every save and commit both the source SASS file as well as the derived CSS file to the repository.

Bottom line is, you don't need to worry about anything in the `src/sass` directory, including Bootstrap. It's compiled to `src/css/main.css`.

I removed everything that is not necessary for Bootstrap to work.

**Comment by nilsvu** [comment]
Concerning you issue with the data structure in MongoDB: I wasn't aware of MongoDB's index functionality until you brought this up. I adjusted the data structure accordingly, thanks for the tip.

File: src/python/WMArchive/PySpark/RecordAggregator.py [hunk 5]
```diff
@@ -14,71 +27,125 @@ def mapper(self, records):
...
-            for site, sdict in hdict.items():
-                print('')
-                print(site)
-                for key, val in sdict.items():
-                    print('%s: %s' % (key, val))
-        return out
+        document = {
+            'stats': {},
+        }
+        for existing_document in records:
```
Comment (vkuznet) [inline src/python/WMArchive/PySpark/RecordAggregator.py:166]: This is code will perform poorly CPU/RAM wise since you're creating a data structure document (which is a dict), then you write it to file and then insert a giganic dict into MongoDB. Instead, I'll suggest to refactor this code and create a function which will yield documents. This generator can be passed to MongoDB client which will insert docs into MongoDB (but you'll need to use proper chunk_size). Let me know if you understand how to refactor this piece.

**Comment by vkuznet** [comment]
Regarding SASS/bootstrap. Could you please clarify the "compiled to src/css/main.css". I understand that sass can reside in src/sass, but what I need to understand which steps (if any) will be required if we'll need to upgrade bootstrap/sass to their newer versions in a future. If some manual "compilation" step is required I want to outline the whole procedure.

**Comment by nilsvu** [comment]
To produce the CSS file from its SASS source, [install SASS](http://sass-lang.com/install) and then run

```
sass src/sass/main.scss src/css/main.css
```

I will also write documentation about it of course.

File: src/python/WMArchive/PySpark/RecordAggregator.py (hunk 5 above)
Comment (nilsvu) [inline src/python/WMArchive/PySpark/RecordAggregator.py:166]: So in my code the mapper fills the `document` dictionaries and the reducer merges them. It's only a dictionary to have an efficient lookup mechanism via the hashed scope. After the merging I grab all of the dictionary's values and insert them into MongoDB. That's now a list of documents, not one huge document anymore.

Since the reducer first needs to do the merging I am not sure how to insert chunks into MongoDB before.

Comment (vkuznet) [inline src/python/WMArchive/PySpark/RecordAggregator.py:166]: ok,
few things first:
- you should not create a new MongoClient in reducer, instead created once in
  constructor of MapReduce class

```
  def __init__(self):
     self.client = MongoClient()
     self.coll = # your collection
     self.chunk_size = 1000 # a chunk size for bulk insert
```
- your reducer takes records so you can pass them into stand-alone function
  which will modify them and yield them back, e.g.

```
def prepare_records(records):
    for existing_document in records:
        # do stuff
        for scope_hash, existing_stats in existing_document['stats'].items():
            # do sutff
            yield stats

```

now modify your reducer method to use prepare_records generator function:

```
import pymongo # to check pymongo version for its APIs
import itertools # to use iterator over generator records
def reducer(self, records, init=0):
    gen = prepare_records(records)
    size = self.cache_size
    try:
        if  pymongo.version.startswith('3.'): # pymongo 3.X
            res = self.coll.insert_many(gen)
            inserted += len(res.inserted_ids)
        else:
            while True:
                nres = self.coll.insert(itertools.islice(gen, size))
                if  nres and isinstance(nres, list):
                    inserted += len(nres)
                else:
                    break
    except InvalidDocument as exp:
       # handle exceptions
```

This code refactoring benefit from:
- has only single MongoClient instance, and reducer will re-use it
- delegate records manipulate into stand-along function (which you can test as you like with any records types)
- generate records via generator and pass generator to mongo collection self.coll
- handle mongo insert exceptions
- don't need to write json files
- work with pymongo 2.X and 3.X APIs to insert bulk of records into collection

Please re-factor code in similar fashion, test it and verify that you insert new
records into MongoDB.

File: src/python/WMArchive/PySpark/RecordAggregator.py [hunk 6]
```diff
@@ -17,10 +18,100 @@ def get_scope_hash(scope):
     return str(hash(frozenset(scope.items())))
 
 
+def extract_stats(record):
```
Comment (vkuznet) [inline src/python/WMArchive/PySpark/RecordAggregator.py:22]: Nils, please add to every function definition its docstring. We'll generate documentation out of docstring. Please be expressive and describe in detail what functions are doing.
